docusafe2-business-impl/src/main/java/de/adorsys/docusafe2/business/impl/dfscredentialservice/DFSCredentialServiceImpl.java

```diff
@@ -0,0 +1,98 @@
+import de.adorsys.docusafe2.business.api.keystore.KeyStoreService;
+import de.adorsys.docusafe2.business.api.keystore.types.*;
+import de.adorsys.docusafe2.business.api.types.DFSCredentials;
+import de.adorsys.docusafe2.business.api.types.DocumentContent;
+import de.adorsys.docusafe2.business.api.types.UserIDAuth;
+import de.adorsys.docusafe2.business.impl.cmsencryption.services.CMSEncryptionServiceImpl;
+import de.adorsys.docusafe2.business.impl.keystore.service.KeyStoreServiceImpl;
+import lombok.extern.slf4j.Slf4j;
+import org.bouncycastle.cms.CMSEnvelopedData;
+import org.bouncycastle.cms.CMSException;
+
+import java.io.IOException;
+import java.security.KeyStore;
+import java.security.PublicKey;
+
+@Slf4j
+public class DFSCredentialServiceImpl implements DFSCredentialService {
+
+    public static final String SYSTEM_DFS_DIRECTORY_NAME = "system-dfs";
+    private static Gson gson = new GsonBuilder().setPrettyPrinting().create();
+
+    private CMSEncryptionService cmsEncryptionService = new CMSEncryptionServiceImpl();
+    private KeyStoreService keyStoreService = new KeyStoreServiceImpl();
+    private DFSConnection dfsConnectionService = DFSConnectionFactory.get();
+
+    @Override
+    public DFSCredentials getDFSCredentials(UserIDAuth userIDAuth) {
+        log.trace("get dfs credential for user: " + userIDAuth.getUserID().getValue());
+
+        KeyStoreAccess keyStoreAccess = getKeyStoreAccess(userIDAuth);
+
+        BucketPath bucketPathForNewUserInSystemDfs = new BucketPath(SYSTEM_DFS_DIRECTORY_NAME, userIDAuth.getUserID().getValue());
+        byte[] data = dfsConnectionService.getBlob(bucketPathForNewUserInSystemDfs).getData();
+        DocumentContent decrypt = null;
+        try {
+            decrypt = cmsEncryptionService.decrypt(new CMSEnvelopedData(data), keyStoreAccess);
+        } catch (CMSException e) {
+            throw new DFSCredentialException(e.getMessage(), e);
+        }
+
+        return gson.fromJson(new String(decrypt.getValue()), DFSCredentials.class);
+    }
+
+    @Override
+    public void registerDFS(DFSCredentials dfsCredentials, UserIDAuth userIDAuth) {
+        log.trace("register dfs for user: " + userIDAuth.getUserID().getValue());
+
+        KeyStoreAccess keyStoreAccess = getKeyStoreAccess(userIDAuth);
+
+        SecretKeyIDWithKey forPublicKey = keyStoreService.getRandomSecretKeyID(keyStoreAccess);
+        KeyID keyID = forPublicKey.getKeyID();
+        PublicKey publicKey = keyStoreService.getPublicKeys(keyStoreAccess).get(0).getPublicKey();
+
+        DocumentContent origMessage = new DocumentContent(gson.toJson(dfsCredentials).getBytes());
+        CMSEnvelopedData encrypted = cmsEncryptionService.encrypt(origMessage, publicKey, keyID);
+
+        BucketPath bucketPathForNewUserInSystemDfs = new BucketPath(SYSTEM_DFS_DIRECTORY_NAME, userIDAuth.getUserID().getValue());
+
+        if(dfsConnectionService.containerExists(bucketPathForNewUserInSystemDfs.getBucketDirectory())) {
```

Comment:
This is correct. Use Exceptions like this. Give as much information as you can provide. Good!